pkg/block/block.go

```diff
@@ -120,29 +126,40 @@ func Upload(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bdir st
 	}
 
 	if meta.Thanos.Labels == nil || len(meta.Thanos.Labels) == 0 {
-		return errors.New("empty external labels are not allowed for Thanos block.")
+		return errors.New("empty external labels are not allowed for Thanos block")
 	}
 
 	meta.Thanos.Files, err = gatherFileStats(bdir, hf, logger)
 	if err != nil {
 		return errors.Wrap(err, "gather meta file stats")
 	}
-
 	metaEncoded := strings.Builder{}
 	if err := meta.Write(&metaEncoded); err != nil {
 		return errors.Wrap(err, "encode meta file")
 	}
 
 	if err := bkt.Upload(ctx, path.Join(DebugMetas, fmt.Sprintf("%s.json", id)), strings.NewReader(metaEncoded.String())); err != nil {
-		return cleanUp(logger, bkt, id, errors.Wrap(err, "upload debug meta file"))
+		if retryCounter == 5 {
+			return cleanUp(logger, bkt, id, errors.Wrap(err, "upload debug meta file"))
+		}
+		flag = true
```

Comment:
Do we really need this flag? We can simply do `return UploadWithRetry(..., retryCounter+1)` here to get the same result.

Reply:
My first commit was exactly that, after which i shifted to this `flag` method. 
Let's say we encounter an error and call `Upload` once again with retry x+1. If that executes successfully we'll be returned back to the scope of function with retry value x.
Actually I feel the flag check should take place in the end, here https://github.com/thanos-io/thanos/blob/fcb62932d5a01c576ea90a33d0efeabc142005ae/pkg/block/block.go#L173 rather than https://github.com/thanos-io/thanos/blob/fcb62932d5a01c576ea90a33d0efeabc142005ae/pkg/block/block.go#L161

Reply:
Rather than using recursive calls for the same, you could create a loop in which you retry for `maxRetryCount` times, in this case, `5`, and if successful, you could break out of the loop. That would be more easier to follow, and you might not need a `flag` variable to do the same.

So something like this:
```go
for i := 0; i < maxRetryCount; i++ {
     if err:= bkt.Upload(...); err == nil {
           break // bucket upload was successful
     }
     log.Debug('retrying again...')
}
```

Reply:
Yes, that makes sense, and instead of the for loop, I'll also go for exponential backoff for each upload